### DISCUSSION TITLE
Button matrix 3 lines 4 colums (3x4)

### DISCUSSION THREAD

**Discussion by thiagovianav** [micropython/micropython#9189] [Hardware & Peripherals]
How i know exactly button as pressed?

**Comment by jimmo** [comment]
hi @ReiLoko4 we're going to need a lot more info please. Do you already have the circuit wired up (or a pre-build keypad module?).

**Reply by thiagovianav** [reply]
Yes, the hardware looks good, but, I don't know how programing this 
3 lines, Pin in 
4 colums, Pin out

**Reply by jimmo** [reply]
COnfigure the three rows as output pins (i.e. `p = machine.Pin(N, mode=machine.Pin.OUT)`, and the four columns as input pins (`mode=machine.Pin.IN`).

You scan your row pins (i.e. turn each one high in turn while the others are low), then for each row, read from each of the column pins.

**Reply by thiagovianav** [reply]
> Yes, the hardware looks good, but, I don't know how programing this 
> 3 lines, Pin in 
> 4 colums, Pin out 

I just know the value in the lines

**Reply by thiagovianav** [reply]
> COnfigure the three rows as output pins (i.e. `p = machine.Pin(N, mode=machine.Pin.OUT)`, and the four columns as input pins (`mode=machine.Pin.IN`).
> 
> 
> 
> You scan your row pins (i.e. turn each one high in turn while the others are low), then for each row, read from each of the column pins.

Its works ? In my board in the rows have resistors and in the colums capacitors 

//sorry for my bad english

**Reply by jimmo** [reply]
Please tell us more info. Maybe a schematic or at least a photo or a link to something with information about how it's wired up.

**Reply by thiagovianav** [reply]
> Please tell us more info. Maybe a schematic or at least a photo or a link to something with information about how it's wired up.

The original project 

![image](https://user-images.githubusercontent.com/103978193/188181144-afdf6e04-aded-469f-89c1-1ff0b4c0a080.jpeg)
![image](https://user-images.githubusercontent.com/103978193/188181251-85b1f8a7-4362-4a40-95e1-c98fbef0155a.jpeg)

L1,l2 and l3 = rows 
C1 etc = colums

**Reply by jimmo** [reply]
>  In my board in the rows have resistors and in the colums capacitors

These are diodes not capacitors.

So you need to scan the columns and read the rows. i.e. turn on 22, 21, 20, 19 one at a time while the rest are off, and then for each, read from 18, 17, 16.

**Comment by LudoJoris** [comment]
Hi,

Can't help you with the Pico, but did a 3x4 button remote controller for the micro:bit.
Here is part of the MicroPython code. It works exactly the way @jimmo replied. 

```
from microbit import *

cols = pin0, pin1, pin2
rows = pin8, pin12, pin14, pin16

cmds = (
  ('+', '0', '-'),
  ('7', '8', '9'),
  ('4', '5', '6'),
  ('1', '2', '3'),
)

for pin in rows + cols:
    pin.write_digital(0)

cmd = None

while True:
    for r, row in enumerate(rows):
        row.write_digital(1)
        for c, col in enumerate(cols):
            if col.read_digital():
                cmd = cmds[r][c]
                break
        row.write_digital(0)
        if cmd:
            break

    if cmd:
        radio.send(cmd)
        cmd = None

```

**Reply by thiagovianav** [reply]
Thx for the help, but I got the result with while 
Turn on energy of coluns and read rows
But turn on just one colun and turn off other 3 (1 on, 3 off)

Sorry for my bad english bro 🥵

**Reply by PsuFan** [reply]
Seems way easier to add a GPIO expander...?

https://github.com/mcauser/micropython-mcp23017
